src/crystal/system/unix/evented/event_loop.cr

```diff
@@ -0,0 +1,506 @@
+require "./*"
+require "./arena"
+
+module Crystal::System::FileDescriptor
+  # user data (generation index for the arena)
+  property __evloop_data : Int64 = -1_i64
+end
+
+module Crystal::System::Socket
+  # user data (generation index for the arena)
+  property __evloop_data : Int64 = -1_i64
+end
+
+module Crystal::Evented
+  # The choice of a generational arena permits to avoid pushing raw pointers into
+  # IO objects into kernel data structures that are unknown to the GC, and to
+  # safely check whether the allocation is still valid before trying to
+  # dereference the pointer. Since `PollDescriptor` also doesn't have pointers to
+  # the actual IO object, it won't prevent the GC from collecting lost IO objects
+  # (and spares us from using
+  #
+  # To a lesser extent, it also allows to keep the `PollDescriptor` allocated
+  # together in the same region, and polluting the IO object itself with specific
+  # evloop data (except for the generation index).
+  #
+  # We assume the fd is unique (OS guarantee) and that the OS will always reuse
+  # the lowest fds before growing, so the memory region should never grow too
+  # big without a good reason (i.e. we need that many fds at that time). This
+  # assumption allows the arena to not have to keep a list of free indexes.
```

Comment:
Switching to dynamically allocated blocks (which I'm doing now) we could hook into the garbage collection, detect which blocks aren't used and free the unused ones (maybe using a red/black strategy) :thinking:

Reply:
AFAIK we can't tell the kernel to deallocate pages: `munmap` will deallocate the pages and further accesses will lead to segfaults; `mremap` is only available on linux and behaves like `realloc`: it may reallocate, invalidating all pointers.

Reply:
AFAIK we can't tell the kernel to deallocate pages: `munmap` will deallocate the pages and further accesses will lead to segfaults; `mremap` is only available on linux and behaves like `realloc`: it may reallocate, invalidating all pointers.

Reply:
Errata: it seems we can use `madvise` with `MADV_DONTNEED` to hint the kernel that some pages aren't needed (can free them if it wants to) yet still allow to access them in the future (at the expense of a page fault).

Linux (4.5+), FreeBSD and more also have `MADV_FREE` which seems to be more proactive to tell the kernel that the pages can be freed, but apparently [it's not working very well](https://github.com/JuliaLang/julia/issues/51086).